src/util/xeval/evaluator.rs

```diff
@@ -386,6 +420,91 @@ impl Evaluator {
         }
     }
 
+    fn eval_json_type(&mut self, ctx: &EvalContext, expr: &Expr) -> Result<Datum> {
+        let child = try!(self.eval_one_child(ctx, expr));
+        if Datum::Null == child {
+            return Ok(Datum::Null);
+        }
+        let json = try!(child.cast_as_json());
+        let json_type = json.json_type().to_vec();
+        Ok(Datum::Bytes(json_type))
+    }
+
+    fn eval_json_merge(&mut self, ctx: &EvalContext, expr: &Expr) -> Result<Datum> {
+        let children = try!(self.eval_more_children(ctx, expr, 2));
+        if children.iter().any(|item| *item == Datum::Null) {
+            return Ok(Datum::Null);
+        }
+        let mut children = children.into_iter();
+        let first = try!(children.next().unwrap().cast_as_json());
+        let suffixes: Vec<Json> = try!(children.map(|item| item.cast_as_json())
+            .collect());
+        Ok(Datum::Json(first.merge(suffixes)))
+    }
+
+    fn eval_json_unquote(&mut self, ctx: &EvalContext, expr: &Expr) -> Result<Datum> {
+        let child = try!(self.eval_one_child(ctx, expr));
+        if child == Datum::Null {
+            return Ok(Datum::Null);
+        }
+        // here Datum::Byte(bs) should be converted into Json::String(bs)
+        // select JSON_UNQUOTE('{"a":   "b"}');
+        // +------------------------------+
+        // | JSON_UNQUOTE('{"a":   "b"}') |
+        // +------------------------------+
+        // | {"a":   "b"}                 |
+        // +------------------------------+
+        let json = try!(child.into_json());
+        let unquote_data = try!(json.unquote());
+        Ok(Datum::Bytes(unquote_data.into_bytes()))
+    }
+
+    fn eval_json_extract(&mut self, ctx: &EvalContext, expr: &Expr) -> Result<Datum> {
+        let children = try!(self.eval_more_children(ctx, expr, 2));
+        if children.iter().any(|item| *item == Datum::Null) {
+            return Ok(Datum::Null);
+        }
+        let mut children = children.into_iter();
+        let json = try!(children.next().unwrap().cast_as_json());
+        let path_extrs: Vec<PathExpression> = try!(children.map(|item| item.as_json_path_expr())
+            .collect());
+        if let Some(data) = json.extract(&path_extrs) {
+            Ok(Datum::Json(data))
+        } else {
+            Ok(Datum::Null)
+        }
+    }
+
+    fn eval_json_modify(&mut self,
+                        ctx: &EvalContext,
+                        expr: &Expr,
+                        mt: ModifyType)
+                        -> Result<Datum> {
+        let children = try!(self.eval_more_children(ctx, expr, 2));
+        if is_even(children.len() as i64) {
+            return Err(Error::Expr(format!("expect odd number operands, got {}", children.len())));
+        }
+
+        if children.iter().any(|item| *item == Datum::Null) {
```

Comment:
Here the logic is wrong. See: https://github.com/pingcap/tidb/pull/3562.
Sorry for that, I made a mistake.